File: codigo/python/interp_newton.py
"""interp_newton.py -- Interpolacion por diferencias divididas de Newton. Chapra cap. 18."""


def coef_newton(x, y):
    """Coeficientes (diferencias divididas) del polinomio de Newton."""
    n = len(x)
    c = y[:]                              # copia
    for j in range(1, n):
        for i in range(n - 1, j - 1, -1):
            c[i] = (c[i] - c[i - 1]) / (x[i] - x[i - j])
    return c


def eval_newton(c, x_datos, xq):
    """Evalua el polinomio de Newton en xq (forma anidada de Horner)."""
    n = len(c)
    p = c[n - 1]
    for k in range(n - 2, -1, -1):
        p = p * (xq - x_datos[k]) + c[k]
    return p


if __name__ == "__main__":
    x = [1, 2, 4, 5]
    y = [0, 1, 2, 3]
    c = coef_newton(x, y)
    for xv in [3, 4.5]:
        print(f"P({xv}) = {eval_newton(c, x, xv):.6f}")

File: codigo/matlab/interp_newton.m
function yq = interp_newton(x, y, xq)
% INTERP_NEWTON  Interpolacion por diferencias divididas de Newton. Chapra cap. 18.
%   Uso:  yq = interp_newton([1 2 4 5], [0 1 2 3], 3);
    n = numel(x); c = y(:);
    for j = 2:n                               % diferencias divididas
        for i = n:-1:j
            c(i) = (c(i) - c(i-1))/(x(i) - x(i-j+1));
        end
    end
    yq = c(n)*ones(size(xq));                 % evaluacion (Horner anidado)
    for k = n-1:-1:1
        yq = yq.*(xq - x(k)) + c(k);
    end
end
